tests/backend/test_bosonic_backend.py

```diff
@@ -0,0 +1,566 @@
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_measurement(self, alpha, r):
+        """Runs a program with measurements."""
+        prog = sf.Program(2)
+        with prog.context as q:
+            sf.ops.Catstate(alpha) | q[0]
+            sf.ops.Squeezed(r) | q[1]
+            sf.ops.MeasureX | q[0]
+            sf.ops.MeasureX | q[1]
+        backend = bosonic.BosonicBackend()
+        applied, samples, all_samples = backend.run_prog(prog)
+        state = backend.state()
+
+        # Check output is vacuum since everything was measured
+        assert np.allclose(state.fidelity_vacuum(), 1)
+
+        # Check samples
+        for i in range(2):
+            assert i in samples.keys()
+            assert samples[i].shape == (1,)
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    @pytest.mark.parametrize("r", R_VALS)
+    def test_mb_gates(self, alpha, r):
+        """Runs a program with measurement-based gates."""
+        prog = sf.Program(2)
+        with prog.context as q:
+            sf.ops.MbSgate(1, 0, 1, 1, avg=True) | q[0]
+            sf.ops.MbSgate(1, 0, 1, 1, avg=False) | q[1]
+            sf.ops.MbSgate(1, 0, 1, 1, avg=False) | q[1]
+        backend = bosonic.BosonicBackend()
+        applied, samples, all_samples = backend.run_prog(prog)
+
+        # Check number of active modes did not change
+        assert backend.get_modes() == [0, 1]
+
+        # Check samples for the data modes are empty
+        assert samples == {}
+
+        # Check ancilla samples exist for the second mode
+        ancilla_samples = backend.ancillae_samples_dict
+        assert 1 in ancilla_samples.keys()
+        assert len(ancilla_samples[1]) == 2
+
+    @pytest.mark.parametrize("alpha", ALPHA_VALS)
+    def test_add_new_mode(self, alpha):
+        """Tests adding a new mode in a program context."""
+        prog = sf.Program(1)
+        with prog.context as q:
+            sf.ops.Vacuum() | q[0]
+            q = q + (sf.ops.New(n=1),)
+            sf.ops.Catstate(alpha) | q[1]
+        backend = bosonic.BosonicBackend()
+        backend.run_prog(prog)
+        state = backend.state()
+        # Check output shapes
+        if alpha != 0:
+            assert state.weights().shape == (4,)
+            assert np.allclose(sum(state.weights()), 1)
+            assert state.means().shape == (4, 4)
+            assert state.covs().shape == (4, 4, 4)
+
+        else:
+            assert state.weights().shape == (1,)
+            assert np.allclose(sum(state.weights()), 1)
+            assert state.means().shape == (1, 4)
+            assert state.covs().shape == (1, 4, 4)
+
+    def test_raised_errors(self):
+        """Runs programs with operations that are not applicable
+        or have not been implemented."""
+        from strawberryfields.backends.base import NotApplicableError
+
+        with pytest.raises(NotApplicableError):
```

Comment:
Suggest reducing the context scope here too.

Reply:
Done.